README.md
# GoTor - HTTP REST API and Web Crawling Tool with TOR Integration

This repository contains an HTTP REST API and a command-line program designed for efficient data gathering and analysis through web crawling using the TOR network. While the program is primarily designed to work seamlessly with TorBot, the API and CLI can also operate independently.

## Status/Social Links
[![Go](https://github.com/DedSecInside/gotor/actions/workflows/go.yml/badge.svg)](https://github.com/DedSecInside/gotor/actions/workflows/go.yml)
[![Open Source Helpers](https://www.codetriage.com/kingakeem/gotor/badges/users.svg)](https://www.codetriage.com/kingakeem/gotor)
[![](https://img.shields.io/badge/Made%20with-Go-blue.svg?style=flat-square)]()
![image](https://github.com/DedSecInside/gotor/assets/13573860/9705fcbf-055c-4024-9f36-1bd4bea71442)

## Features and Options

### TOR Integration
The program employs the TOR network for enhanced privacy and security during web crawling. TOR settings can be configured using environment variables or overridden using CLI flags.

* `-h`: Specify the SOCKS5 proxy host (default: localhost / 127.0.0.1)
* `-p`: Specify the SOCKS5 proxy port (default: 9050)

### REST API
The REST API is accessible by default at http://localhost:8081 and provides the following endpoint, check the `api` directory for endpoint documentation.
  
### Available Crawling Mechanisms
1. **Building Relationship Tree of Links**: Generates a hierarchical tree of links, with child nodes representing links found on a website.
2. **Getting Tor Client IP**: Retrieves the IP address of the current TOR client.
3. **Retrieving Phone Numbers**: Collects phone numbers found on websites.
4. **Retrieving Emails**: Gathers email addresses found on websites.

#### Example Usage
To start the HTTP server and initiate crawling, use the following command:
```bash
go run cmd/main/main.go -server
```

w/ alternate host and port:
```bash
go run cmd/main/main.go -server -h localhost -p 9051
```

To crawl directly using the CLI and output the results to an Excel file, use the following command:
```bash
go run cmd/main/main.go -l https://example.com -d 2 -o excel
```

## Running with Docker
To run the server using Docker, a convenience script build.sh is provided. This script builds a Docker network service for Tor and connects it to the "gotor" Docker container.
Make sure no other service is using the same port. The script uses the SOCKS5_PORT.

### To build and start the Docker containers:
```bash
./build.sh
```
### To stop and destroy the Docker containers:
```bash
./destroy.sh
```

### Additional Configuration
* `-l`: the URL to use if searching directly
* `-d`: Specify the depth for searching children nodes of links (default: 1)
* `-o`: Set the output destination for results. Supported formats include:
	* terminal (outputs the tree structure directly to the terminal)
	* excel (saves results to a .xlsx file in the current directory)
	* json (saves results to a .json file in the current directory)
 
To disable TOR, set the USE_TOR flag in the .env file.

## Documentation
This project includes comprehensive code comments to facilitate documentation generation with godoc. To generate and access documentation, use the following command:

```bash
godoc -v -http=:6060
```
This will make the documentation available at http://127.0.0.1:6060.

## License
This project is licensed under the GNU General Public License.

Feel free to contribute, report issues, or suggest improvements!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v2.0.0 - 2023-08-28
Bug fixes, slight performance updates, and some refactoring

## What's Changed
* Add logging by @KingAkeem in https://github.com/DedSecInside/gotor/pull/35
* Add configuration package and .env file, utilize env variables by @KingAkeem in https://github.com/DedSecInside/gotor/pull/37
* Add docker by @KingAkeem in https://github.com/DedSecInside/gotor/pull/39
* Changed `USE_TOR` default value to `true`. by @fukusuket in https://github.com/DedSecInside/gotor/pull/40
* Refactoring and improving documentation by @KingAkeem in https://github.com/DedSecInside/gotor/pull/41
* Improving docker by @KingAkeem in https://github.com/DedSecInside/gotor/pull/42
* Bump golang.org/x/net from 0.4.0 to 0.7.0 by @dependabot in https://github.com/DedSecInside/gotor/pull/47
* Update README.md by @KingAkeem in https://github.com/DedSecInside/gotor/pull/48

## New Contributors
* @fukusuket made their first contribution in https://github.com/DedSecInside/gotor/pull/40
* @dependabot made their first contribution in https://github.com/DedSecInside/gotor/pull/47

**Full Changelog**: https://github.com/DedSecInside/gotor/compare/v1.0.1...v2.0.0

## v1.0.1 - 2022-11-10

### What's Changed
* Feature #22: Crawl phone numbers by @PSNAppz in https://github.com/DedSecInside/gotor/pull/25
* Add struct tags by @KingAkeem in https://github.com/DedSecInside/gotor/pull/26
* Patch: Get phone number if exists by @PSNAppz in https://github.com/DedSecInside/gotor/pull/27
* Get Web Content API by @PSNAppz in https://github.com/DedSecInside/gotor/pull/28
* Reorganize code and update documentation by @KingAkeem in https://github.com/DedSecInside/gotor/pull/29

### New Contributors
* @PSNAppz made their first contribution in https://github.com/DedSecInside/gotor/pull/25

**Full Changelog**: https://github.com/DedSecInside/gotor/compare/v1.0.0...v1.0.1

## v1.0.0 - 2021-09-28

### What's in this release?

- Resolves issues with depth argument when crawling a node/building a tree
- Unit tests for tree logic
- Create token stream pipeline


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strconv"

	"github.com/KingAkeem/gotor/api"
	"github.com/KingAkeem/gotor/linktree"
	"github.com/gorilla/mux"
	"github.com/mgutz/ansi"
	"github.com/xuri/excelize/v2"
)

func logInfo(msg string) {
	log.Println(ansi.Color(msg, "blue"))
}

func logErr(err error) {
	log.Println(ansi.Color(err.Error(), "red"))
}

// creates a http client using socks5 proxy
func newTorClient(host, port string) (*http.Client, error) {
	proxyStr := fmt.Sprintf("socks5://%s:%s", host, port)
	proxyURL, err := url.Parse(proxyStr)
	if err != nil {
		return nil, err
	}
	transport := &http.Transport{
		Proxy: http.ProxyURL(proxyURL),
	}
	return &http.Client{
		Transport: transport,
	}, nil
}

func writeTree(node *linktree.Node, depth int) {
	node.Load(depth)
	node.PrintTree()
}

func writeTerminal(node *linktree.Node, depth int) {
	printStatus := func(link string) {
		n := linktree.NewNode(node.Client, link)
		markError := ansi.ColorFunc("red")
		markSuccess := ansi.ColorFunc("green")
		if n.StatusCode != 200 {
			fmt.Printf("Link: %20s Status: %d %s\n", n.URL, n.StatusCode, markError(n.Status))
		} else {
			fmt.Printf("Link: %20s Status: %d %s\n", n.URL, n.StatusCode, markSuccess(n.Status))
		}
	}
	node.Crawl(depth, printStatus)
}

func writeExcel(node *linktree.Node, depth int) {
	f := excelize.NewFile()
	err := f.SetCellStr(f.GetSheetName(0), "A1", "Link")
	if err != nil {
		log.Fatal(err)
		return
	}
	err = f.SetCellStr(f.GetSheetName(0), "B1", "Status")
	if err != nil {
		log.Fatal(err)
		return
	}
	row := 2
	addRow := func(link string) {
		node := linktree.NewNode(node.Client, link)
		linkCell := fmt.Sprintf("A%d", row)
		statusCell := fmt.Sprintf("B%d", row)
		err = f.SetCellStr(f.GetSheetName(0), linkCell, node.URL)
		if err != nil {
			log.Fatal(err)
			return
		}
		err = f.SetCellStr(f.GetSheetName(0), statusCell, fmt.Sprintf("%d %s", node.StatusCode, node.Status))
		if err != nil {
			log.Fatal(err)
			return
		}
		row++
	}
	node.Crawl(depth, addRow)
	u, err := url.Parse(node.URL)
	if err != nil {
		log.Fatal(err)
		return
	}
	filename := fmt.Sprintf("%s_depth_%d.xlsx", u.Hostname(), depth)
	err = f.SaveAs(filename)
	if err != nil {
		log.Fatal(err)
		return
	}
}

func runServer(host, port string) {
	router := mux.NewRouter()

	client, err := newTorClient(host, port)
	if err != nil {
		log.Fatal(err)
		return
	}

	router.HandleFunc("/ip", api.GetIP(client)).Methods(http.MethodGet)
	router.HandleFunc("/emails", api.GetEmails(client)).Methods(http.MethodGet)
	router.HandleFunc("/phone", api.GetPhoneNumbers(client)).Methods(http.MethodGet)
	router.HandleFunc("/tree", api.GetTreeNode(client)).Methods(http.MethodGet)
	router.HandleFunc("/content", api.GetWebsiteContent(client)).Methods(http.MethodGet)

	logInfo("Listening on port 8081")
	err = http.ListenAndServe(":8081", router)
	if err != nil {
		logErr(err)
		return
	}
}

func main() {
	var root string
	var host string
	var port string
	var depthInput string
	var output string
	var serve bool
	flag.StringVar(&root, "l", "", "Root used for searching. Required. (Must be a valid URL)")
	flag.StringVar(&depthInput, "d", "1", "Depth of search. Defaults to 1. (Must be an integer)")
	flag.StringVar(&host, "h", "127.0.0.1", "The host used for the SOCKS5 proxy. Defaults to localhost (127.0.0.1.)")
	flag.StringVar(&port, "p", "9050", "The port used for the SOCKS5 proxy. Defaults to 9050.")
	flag.StringVar(&output, "o", "terminal", "The method of output being used. Defaults to terminal.")
	flag.BoolVar(&serve, "server", false, "Determines if the program will behave as an HTTP server.")
	flag.Parse()

	// If the server flag is passed then all other flags are ignored.
	if serve {
		runServer(host, port)
		return
	}

	if root == "" {
		flag.CommandLine.Usage()
		return
	}

	depth, err := strconv.Atoi(depthInput)
	if err != nil {
		flag.CommandLine.Usage()
		return
	}

	client, err := newTorClient(host, port)
	if err != nil {
		log.Fatal(err)
		return
	}

	node := linktree.NewNode(client, root)
	switch output {
	case "terminal":
		writeTerminal(node, depth)
	case "excel":
		writeExcel(node, depth)
	case "tree":
		writeTree(node, depth)
	}
}
